### PR TITLE
Add infer request queue for tokenizers and allow for optional plugin_config in tokenizer

### PR DESCRIPTION
**Pull request by dkalinowski**
This improves performance of CB lib when tested within OVMS.

### CODE REVIEW

File: src/cpp/src/tokenizer.cpp [hunk 1]
```diff
@@ -91,11 +96,25 @@ class Tokenizer::TokenizerImpl {
...
-                                                device).create_infer_request();
-        m_detokenizer_request = core.compile_model(tokenizer_path / "openvino_detokenizer.xml", 
-                                                   device).create_infer_request();
+    
+        m_tokenizer = core.compile_model(tokenizer_path / "openvino_tokenizer.xml",
+                                                device, plugin_config);
+        m_detokenizer = core.compile_model(tokenizer_path / "openvino_detokenizer.xml", 
+                                                   device, plugin_config);
+
+        const size_t INFER_REQUEST_QUEUE_SIZE = 256;
```
Comment (dtrawins) [inline src/cpp/src/tokenizer.cpp:107]: I suggest to read the compile_model property NUM_STREAMS and set the queue size to such value

File: src/cpp/src/tokenizer.cpp [hunk 2]
```diff
@@ -260,22 +283,22 @@ class Tokenizer::TokenizerImpl {
...
     std::vector<std::string> decode(ov::Tensor tokens) {
         OPENVINO_ASSERT(tokens.get_element_type() == ov::element::i64, "tokens tensor element type should be an i64");
         OPENVINO_ASSERT(tokens.get_shape().size() == 2, "tokens tensor should of rank 2 with shape [batch_size, seq_len]");
 
-        std::unique_lock<std::mutex> lock(m_detokenizer_mutex);
-        m_detokenizer_request.set_input_tensor(tokens);
-        m_detokenizer_request.infer();
+        CircularBufferQueueElementGuard<ov::InferRequest> infer_request_guard(this->m_ireq_queue_detokenizer.get());
+        infer_request_guard.get().set_input_tensor(tokens);
+        infer_request_guard.get().infer();
```
Comment (dtrawins) [inline src/cpp/src/tokenizer.cpp:299]: wouldn't it be better to use async call and wait for result? It used to be more efficient execution with multiconcurrency.

File: src/cpp/src/tokenizer.cpp [hunk 3]
```diff
@@ -91,11 +96,25 @@ class Tokenizer::TokenizerImpl {
         // Try to read tokenizer_config if some token ids or token str are not defined.
         read_tokenizer_config_if_necessary(tokenizer_path); 
 
+        std::map<std::string, ov::Any> plugin_config{{"PERFORMANCE_HINT", "THROUGHPUT"}};
```
Comment (Wovchena) [inline src/cpp/src/tokenizer.cpp:99]: My expectation was that tokenizers are single threaded. @apaniukov, does this affect anything?

File: src/cpp/src/tokenizer.cpp (hunk 1 above)
Comment (Wovchena) [reply src/cpp/src/tokenizer.cpp:107]: Good idea. There's optimal_number_of_infer_requests. Read it instead. It's meta plugin aware

File: src/cpp/src/tokenizer.cpp (hunk 2 above)
Comment (Wovchena) [reply src/cpp/src/tokenizer.cpp:299]: Why would it help? encode() is still a blocking call.

File: src/cpp/src/tokenizer.cpp (hunk 3 above)
Comment (apaniukov) [reply src/cpp/src/tokenizer.cpp:99]: Yes, the throughput hint massively increases the async inference scenario. For sync inference, a latency hint is better.

Comment (Wovchena) [reply src/cpp/src/tokenizer.cpp:99]: That means synchronous scenario becomes slower with that change?

Comment (apaniukov) [reply src/cpp/src/tokenizer.cpp:99]: Yes, the average latency is a bit bigger in that scenario. There is a benchmarking script that measures sync and async inference stats for a particular tokenizer: https://github.com/openvinotoolkit/openvino_tokenizers/tree/master/benchmark
You can pass `--tput` flag to activate the throughput hint for the benchmark.

File: src/cpp/src/tokenizer.cpp (hunk 1 above)
Comment (dkalinowski) [reply src/cpp/src/tokenizer.cpp]: Done

File: src/cpp/src/tokenizer.cpp (hunk 3 above)
Comment (dkalinowski) [reply src/cpp/src/tokenizer.cpp]: To enable latency and tput scenarios I have added tokenizer plugin config as constructor argument. This way existing workflows will not be modified, but in OVMS we will manually set tput mode.

Please re-review.

File: src/cpp/src/tokenizer.cpp (hunk 2 above)
Comment (dkalinowski) [reply src/cpp/src/tokenizer.cpp]: I don't think we ever investigated why, but our internal OVMS perf tests showed throughput increase when using `start_async`+`wait` instead of `infer`.

Changed to start_async+wait @dtrawins . Please let me know what you think @Wovchena

File: src/cpp/src/tokenizer.cpp (hunk 1 above)
Comment (Wovchena) [reply src/cpp/src/tokenizer.cpp]: You still read `num_streams`: `const size_t INFER_REQUEST_QUEUE_SIZE = m_tokenizer.get_property(ov::num_streams);`

Comment (dkalinowski) [reply src/cpp/src/tokenizer.cpp]: Right, I have changed it to `optimal_number_of_infer_requests` now

File: src/cpp/include/openvino/genai/continuous_batching_pipeline.hpp [hunk 4]
```diff
@@ -30,7 +30,8 @@ class OPENVINO_GENAI_EXPORTS ContinuousBatchingPipeline {
     ContinuousBatchingPipeline(const std::string& models_path,
                                const SchedulerConfig& scheduler_config,
                                const std::string& device = "CPU",
-                               const ov::AnyMap& plugin_config = {});
+                               const ov::AnyMap& llm_plugin_config = {},
+                               const ov::AnyMap& tokenizer_plugin_config = {});
```
Comment (Wovchena) [inline src/cpp/include/openvino/genai/continuous_batching_pipeline.hpp:34]: @apaniukov just reminded. Please update python bindings

Comment (Wovchena) [reply src/cpp/include/openvino/genai/continuous_batching_pipeline.hpp:34]: You can also minimize the chance of us breaking `tokenizer_plugin_config` and Tokenizer's `plugin_config` args by adding tests. :) But since you are the only users of that feature it's up to you to write the tests.

Comment (dkalinowski) [reply src/cpp/include/openvino/genai/continuous_batching_pipeline.hpp:34]: Added the bindings, rebased and resolved conflicts with latest master.
I did not add new tests, however I explicitly pass empty plugin config `{}` to Tokenizer/CB pipeline in tests, this way we should have ensured that nobody changes the constructor without modifying those tests/fixtures.

Please re-review @Wovchena

**Review by Wovchena — APPROVED** [review]

**Review by dtrawins — APPROVED** [review]

File: src/cpp/src/circular_buffer_queue.hpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
+#pragma once
```
Comment (ilya-lavrenov) [inline src/cpp/src/circular_buffer_queue.hpp:1]: could you please add copy right header?

File: tests/python_tests/common.py [hunk 6]
```diff
@@ -273,7 +273,7 @@ def run_continuous_batching(
     prompts: List[str],
     generation_configs : List[GenerationConfig]
 ) -> List[GenerationResult]:
-    pipe = ContinuousBatchingPipeline(model_path.absolute().as_posix(), scheduler_config, "CPU", {})
+    pipe = ContinuousBatchingPipeline(model_path.absolute().as_posix(), scheduler_config, "CPU", {}, {})
```
Comment (ilya-lavrenov) [inline tests/python_tests/common.py:276]: why do we need to pass empty brackets? As I see we have a default value for this parameter

Comment (Wovchena) [reply tests/python_tests/common.py:276]: The explanation is in https://github.com/openvinotoolkit/openvino.genai/pull/651#discussion_r1687871150
